### PR TITLE
raft: Make flow control more aggressive

### PR DESCRIPTION
**Pull request by bdarnell**
We allow multiple in-flight append messages, but prior to this change
the only way we'd ever send them is if there is a steady stream of new
proposals. Catching up a follower that is far behind would be
unnecessarily slow (this is exacerbated by a quirk of CockroachDB's
use of raft which limits our ability to catch up via snapshot in some
cases).

See cockroachdb/cockroach#27983

Please read https://github.com/coreos/etcd/blob/master/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: raft/raft.go [hunk 1]
```diff
@@ -441,30 +441,37 @@ func (r *raft) getProgress(id uint64) *Progress {
...
 }
 
-// sendAppend sends RPC, with entries to the given peer.
-func (r *raft) sendAppend(to uint64) {
+// sendAppend sends an append RPC with new entries to the given peer,
+// if necessary. Returns true if a message was sent. The sendIfEmpty
+// argument controls whether messages with no entries will be sent
+// ("empty" messages are useful to convey updated Commit indexes, but
+// are undesirable when we're sending multiple messages in a batch).
+func (r *raft) sendAppend(to uint64, sendIfEmpty bool) bool {
```
Comment (xiang90) [inline raft/raft.go]: can we keep the old sendAppend? and add a new func maybeSendAppend (or sendAppendIfNotEmpty)?

**Comment by xiang90** [comment]
@bdarnell Have you tried this in CockroachDB setup?

**Comment by xiang90** [comment]
LGTM after fixing the nit.

**Comment by siddontang** [comment]
> Yes, we've tested it. We had a node that had accumulated a multi-gigabyte raft log. Without this change it was going to take weeks to catch up, and with this it caught up in hours.

Hi @benbjohnson 

Why don't you use snapshot here? seem your Range size is 64 MB, the total log size has exceeded this threshold too much.

**Pull request merged**